# Python WorkShop

### 目次<br>

- Python基本文法<br>
    - 基本演算
    - リスト, タプル, 辞書
    - if, for, while文
    - 関数
    - クラス
    - エラー文
- Pythonの有名ライブラリ<br>
    - numpy(scipy), collections
    - matplotlib
- Python環境構築<br>
    - 仮想環境(venv)

- 扱わないこと
<br>Pythonのビット演算, sklearnなどの機械学習ライブラリ(時間の関係で扱いません、<br>でもsklearnなどのライブラリをただ使ってきただけの人間には刺さる内容にしているはずです)

#### Python基本文法

#### 基本演算
代入
```python
x = 3
```
と書くことでxには値が変更されない限り3という値が入っています。<br>
printという関数で中身を確認することができます。<br>
難しい話になりますがこの時の出力先は標準出力と呼ばれています。<br>

In [1]:
x = 3
print(x)

3


数値型<br>
Pythonではint, float, complex(大まかに整数, 実数, 虚数)という数字の型が用意されています。<br>
type関数でそれぞれの数の型を確認することができます。<br>
また、コードに関係ない部分は「#」を置くことで一行のコメントアウトすることが出来、<br>
「"""」で複数行にわたってコメントアウトが可能になります<br>

In [2]:
x = 3
y = 10.1
z = 1+2j #複素数の場合はiではなくjで複素数化することに注意
print(x)
print(y)
print(z)
print(type(x))
print(type(y))
print(type(z))

3
10.1
(1+2j)
<class 'int'>
<class 'float'>
<class 'complex'>


数値型の代表的な演算として+, -, *, /, //, %, **があげられます。+, -は数学で使われる記号と同じです。<br>
*は掛け算を計算します。<br>
/は割り算を計算します。答えを小数まで出力します<br>
//は割り算の商を計算します。<br>
%は割り算の余りを計算します。<br>
**は乗数を計算します<br>

In [3]:
x = 10
y = 3
print(x + y) #10 + 3
print(x - y) #10 - 3
print(x * y) #10 * 3
print(x / y)#10 / 3
print(x // y)#10 // 3
print(x % y)#10を3で割った余り
print(x ** y)# 10^3

13
7
30
3.3333333333333335
3
1
1000


10 / 3の答えが本当は1/3≒0.3333333....ですが今回は3.3333333333333335となっています。<br>
これはコンピューターが表現できる数値の限界の精度を超えたために生じる誤差になっています。<br>
詳しくは[コンピュータで扱う数値の誤差について](https://www.foresight.jp/fe/column/about-error/)を参照してください

文字列<br>
「'」もしくは「"」で囲んだものは文字列として解釈されます<br>
文字列型はprint関数と組み合わせることで標準出力に日本語や英語を書くことができます。<br>
文字列は足し算が定義でき、文字列の長さはlen関数で取得ができます

In [4]:
string1 = "Hello"
string2 = "World"
print(string1) #Helloと出力される
print(string1 + string2) #文字列の足し算は文字列の結合となる
print(string1 + str(2023)) #str関数で数値などを文字列に変換することができる
print(len(string1)) #string1の文字列の長さがわかる

Hello
HelloWorld
Hello2023
5


文字列のフォーマット<br>
f"{a}, {b}"、もしくは"{}, {}".format(a, b)とすることで、<br>
文字列ではないものを文字列として解釈させることが可能になる。<br>
個人的にはミスが減るので前者推奨

In [5]:
a = 2023
b = 2024
print(f"{a}-{b}")
print("{}-{}".format(a, b))

2023-2024
2023-2024


真偽値(bool)<br>
ある条件文に対してその条件が正しい場合にはTrue, 間違っている場合にはFalseを出力する<br>
条件文などに使われる演算子として「>」, 「<」, 「==」などがあげられ<br>
「>」や「<」は不等号を表し、「1>3」は1が3より大きいかを判定する演算子である。<br>
「==」は等号を表し、「1==3」は1と3が等しいかを判定する演算子である<br>

In [6]:
a = True
print(a) 
print(1 > 3) #1は3より小さいのでFalseが代入される
print(1 == 3)
a = 1 == 1 #左の文はaに1==1の結果が代入されることに注意
print(a)

True
False
False
True


真偽値についての詳しい解説は[Pythonの真偽値bool型（True, False）と他の型との変換・判定](https://note.nkmk.me/python-bool-true-false-usage/)を参照

#### リスト, タプル, 辞書

リスト<br>
リストは[1,2,3,4]のように[]の中に要素を配置する<br>
Pythonでは他の言語(例えばC言語のように配列の要素数をあらかじめ指定する必要はない)<br>

In [7]:
a = [1,2,3,4]
print(a[0]) #要素のアクセスは List[index]のようにおこなうこの時, 
            #最初の要素は0, 二番目の要素は1, ...のように0始まりのindexであることに注意
print(a[0:2]) # start_index:end_indexでstart_indexからend_index-1までの要素を出してくれる。この場合index = 0, 1が出る。
print(a[0:4:2]) # start_index:end_index:stepでstep飛ばしで数を出力する。この場合index = 0, 2の値が出る
print(a[-1]) #indexには負の数値も使える。-1とすると,リストの一番末尾のものを取り出すことを意味する
print(a[::-1]) #覚えておくと良い記法, リストの順番を逆にするときはList[::-1]
a.append(5)
print(a) #要素はappendで追加される
del a[4] #要素はdelで消す
b = [5, 6]
a.append(b)
print(a) #appendはただ単純に要素を追加する
del a[4] #要素はdelで消す
a.extend(b) #extendはリストを入れるとリストの中身を展開する
print(a)
del a[5], a[4]
print(a + b) #リストは足し算をするとextendと似た挙動を示す。
print(len(a)) #リストの要素数はlen関数で取得する

1
[1, 2]
[1, 3]
4
[4, 3, 2, 1]
[1, 2, 3, 4, 5]
[1, 2, 3, 4, [5, 6]]
[1, 2, 3, 4, 5, 6]
[1, 2, 3, 4, 5, 6]
4


リストを使う際の注意点<br>
あるリスト一つを作り, そのリストの要素を変更する前と変更した後でどうなるかを見てみよう<br>
この時、オブジェクトのidを確認するid関数でidを見てみよう

In [8]:
a = [1,2,3,4]
b = a
print(f"List a: {a}, List b: {b}")
print(f"List a's id: {id(a)}, List b's id: {id(b)}")
b[0] = 6 #リストbの要素を変更
print(f"List a: {a}, List b: {b}") #aの要素が変化してしまっている
print(f"List a's id: {id(a)}, List b's id: {id(b)}")

List a: [1, 2, 3, 4], List b: [1, 2, 3, 4]
List a's id: 4361946688, List b's id: 4361946688
List a: [6, 2, 3, 4], List b: [6, 2, 3, 4]
List a's id: 4361946688, List b's id: 4361946688


リストbの要素の変化がリストaの要素に反映されているのはリストaとリストbのidが同一であるからと考えることができる。<br>
これはC系の言語におけるポインタをb=aでは代入しているためである。<br>
では整数はどうなのだろうか

In [9]:
a = 5
b = a
print(f"int a: {a}, int b: {b}")
print(f"int a's id: {id(a)}, int b's id: {id(b)}, 5's id: {id(5)}")
b = 10
print(f"int a: {a}, int b: {b}")
print(f"int a's id: {id(a)}, int b's id: {id(b)}, 10's id: {id(10)}")

int a: 5, int b: 5
int a's id: 4308606024, int b's id: 4308606024, 5's id: 4308606024
int a: 5, int b: 10
int a's id: 4308606024, int b's id: 4308606184, 10's id: 4308606184


Pythonではint, str, boolのオブジェクトはimmutableなオブジェクトと呼ばれ、固有のid値を持っている<br>
このため数値ではb=aのような代入をした後にbの値を変化してもaの値は変化しない<br>
immutableやmutableの概念に関しては別の文献を検索して調べてみてほしい

リストの受け渡し方法は何通りかある。

In [10]:
import copy #copyを使った方法
a = [1,2,3,4]
b = a[:] #スライス記法を使った方法
print(f"List a: {a}, List b: {b}")
print(f"List a's id: {id(a)}, List b's id: {id(b)}")
b = copy.copy(a)
print(f"List a: {a}, List b: {b}")
print(f"List a's id: {id(a)}, List b's id: {id(b)}")
b = copy.deepcopy(a)
print(f"List a: {a}, List b: {b}")
print(f"List a's id: {id(a)}, List b's id: {id(b)}")

List a: [1, 2, 3, 4], List b: [1, 2, 3, 4]
List a's id: 4359765312, List b's id: 4359764928
List a: [1, 2, 3, 4], List b: [1, 2, 3, 4]
List a's id: 4359765312, List b's id: 4359764544
List a: [1, 2, 3, 4], List b: [1, 2, 3, 4]
List a's id: 4359765312, List b's id: 4359765568


copyとdeepcopyの違いは主に二次元配列やクラスを扱うときに重要となるが、今回は割愛する

タプル<br>
上で見たようにリストは値の変更に大きく影響を受ける。そのため、値の変更に影響を受けないものをタプルといい、<br>
()の中に要素を配置して制作する

In [11]:
a = (1,2,3,4)
print(a)
print(len(a))

(1, 2, 3, 4)
4


要素を変更しようとすると以下のようなエラーを吐く
```python
a = (1,2,3,4)
a[0] = 2
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[53], line 2
      1 a = (1,2,3,4)
----> 2 a[0] = 2

TypeError: 'tuple' object does not support item assignment
```

辞書<br>
辞書は{"Bob":16, "Mary":18, "Ken": 17}のように{key: value}のように{}で囲んで定義する<br>

In [21]:
dictionary = {"Bob":16, "Mary":18, "Ken": 17}
print(dictionary["Bob"]) #dict[key]でそのkeyのvalueを確認できる
print(dictionary.keys()) #.keysメソッドでkeyを確認できる
print(dictionary.values()) #.valuesメソッドでvalueを確認できる
print(dictionary.items()) #.itemsで(key, value)のtupleで辞書の中身を確認できる
#上で述べたkeys, values, itemsは後に解説するfor文で中身を取り出すときに使われる

16
dict_keys(['Bob', 'Mary', 'Ken'])
dict_values([16, 18, 17])
dict_items([('Bob', 16), ('Mary', 18), ('Ken', 17)])


if文<br>
if文は主に以下の形で使われる
```python
a = True
b = False
if a:
    print("a is True.")
elif b:
    print("a is False, but b is True.")
else:
    print("Both are False.")
```
まずif (条件文):の形で条件文の中身がTrueかを判断する, もしFalseならelifに移動し、elif (条件文):の条件文がTrueかどうかを判断する<br>
もしFalseならelseに移動し、elseの中身を実行する。

In [12]:
a = True
b = False
if a:
    print("a is True.") #この場合はaがTrueなのでここが実行される
elif b:
    print("a is False, but b is True.")
else:
    print("Both are False.")

a is True.


if, for, while文の中身では改行(インデント)をして半角スペースを置いたところからコードを書き始めます。<br>
このインデントで仕切られた区間をブロックをいいます。<br>
例えば、以下のように改行してないコードなどはエラーとなります。<br>
半角スペースはVSCodeなどでは自動で調整してくれますが、Tabキーを押すことで半角4つ分のスペースを置くことができます。

複数の条件を指定するときにはand, orなどが使えて、否定条件はnotが使えます。

In [2]:
a = True
b = False
if a and b:
    print("ここは実行されない")
else:
    print("aとbは共にTrueではないのでここが実行される")
if a or b:
    print("aがTrueなのでここが実行される")
else:
    print("ここは実行されない")
if not b:
    print("bはFalseなのでここが実行される")
else:
    print("ここは実行されない")

aとbは共にTrueではないのでここが実行される
aがTrueなのでここが実行される
bはFalseなのでここが実行される


```python
a = True
b = False
if a:
print("a is True.")
  Cell In[12], line 4
    print("a is True.") #この場合はaがTrueなのでここが実行される
    ^
IndentationError: expected an indented block after 'if' statement on line 3
```

for文<br>
for文はfor A in B: という形で使われる。
```python
for A in B:
    (処理)
```
最も多い形はrange関数を使うものであり、<br>
```python
for i in range(n):
    (処理)
```
のようなものである。まずは書いてみて挙動を抑えよう

In [24]:
n = 4
print("range(n)を使ったfor文")
for i in range(n):
    print(i)
l = [0,1,2,3]
print("Listを使ったfor文")
for i in l:
    print(i)
print("辞書を使ったfor文")
dictionary = {"Bob":16, "Mary":18, "Ken": 17}
for key in dictionary:
    print(key) #辞書はkeyが取り出される
print("辞書の.items()を使ったfor文")
for key, value in dictionary.items():
    print(key, value) #辞書はkeyが取り出される (key, value)というtupleをfor文では取り出しているが、forの前に2つ変数を置くことでtupleの2つの変数を代入している

range(n)を使ったfor文
0
1
2
3
Listを使ったfor文
0
1
2
3
辞書を使ったfor文
Bob
Mary
Ken
辞書の.items()を使ったfor文
Bob 16
Mary 18
Ken 17


In [15]:
#もう少しrangeについて補足
#range(n)は0からn-1までの数を取得するが、range(start, end)はstartからend-1までの数を取得する
print("range(4)の出力")
for i in range(4):
    print(i)
print("range(1, 4+1)の出力")
for i in range(1, 4+1):
    print(i)
#range(start, end, step)でstartからend-1までstep飛ばしで出力する。stepには負の数を指定することも可能
print("range(1, 4+1, 2)の出力")
for i in range(1, 4+1, 2):
    print(i)
print("range(4,1-1,-1)の出力")
for i in range(4,1-1,-1):
    print(i)

range(4)の出力
0
1
2
3
range(1, 4+1)の出力
1
2
3
4
range(1, 4+1, 2)の出力
1
3
range(4,1-1,-1)の出力
4
3
2
1


リスト内包表記<br>
Pythonのコードでかなりみる書き方。例えばList = [1,2,3,4,5]という配列があったときその要素を二乗した要素を作りたいという状況を考える。<br>
このとき、[x ** 2 for x in List]と書くと、Listを先頭から1要素ずつ読み出して、要素を二乗して新しいリストの要素にするという意味になる。

In [26]:
List = [1,2,3,4,5]
print([x ** 2 for x in List]) #リスト内包表記で作られたリストをprint

[1, 4, 9, 16, 25]


while文<br>
while文はwhile (条件文):のあとに処理を書く
```python
while (条件文):
    (処理)
```
while文は条件がTrueであり続ける限り, 実行され続ける。そのため、条件がなりたたなくなるような処理を絶対に書かなくてはならない

In [1]:
i = 5
while i > 0:
    print(i)
    i -= 1

5
4
3
2
1


条件文のところにリストを置くこともできる。この場合、リストに要素があり続ける限り実行され続ける。<br>
popでリストから要素を取り出していった時の挙動を確認しよう

In [3]:
l = [1,2,3,4,5]
while l:
    a = l.pop()
    print(a)

5
4
3
2
1


#### 関数

関数とは複数行に跨る処理を一気にやってしまおうとするものである。<br>
関数はpythonでは以下のように定義する<br>
```Python
def (関数名):
    (処理)
    return (返り値)
```
簡単に1からnまでの和を求めるプログラムをここでは書いてみよう

In [10]:
def cumsum(n):
    cumulative_sum = 0
    for i in range(1, n+1):
        cumulative_sum += i
    return cumulative_sum
print(cumsum(10))

55


上で定義された関数cumsumの関数の実行の際にいれているnを引数と呼びます。<br>
returnされる値を関数の返り値と呼びます。

キーワード引数<br>
引数の対応を明示的に指定して実行することもできます

In [1]:
def power(x, y):
    return x ** y #xのy乗を計算する
print(power(x = 2, y = 3))
print(power(y = 3, x = 2))

8
8


デフォルト引数<br>
引数を指定しなかった時に自動で代入される引数を決めることができます<br>
下のpower関数はyが代入されなかった場合にはxの1乗を計算する

In [2]:
def power(x, y = 1):
    return x ** y #xのy乗を計算する
print(power(2))

2


型ヒント<br>
Pythonに型はないが、githubなどでは型が明示的に書いてあるコードが見れる。<br>
ある数nが素数かいなかを判定する関数isPrimeでこの型ヒントを学んでみよう<br>
基本形は以下の通りである。<br>
```python
def function(x: type, y: type)-> return type:
```

In [8]:
def isPrime(n: int)->bool:
    if n == 1:
        return False
    for i in range(2, n):
        if n % i == 0:
            return False
    return True
print(isPrime(5))

True


再帰関数<br>
再帰関数は関数の中で関数を実行させて処理を行う関数である。<br>
再帰関数は必ず止まるように設計しなければならない<br>
```python
def function(引数):
    if (停止条件):
        return (停止条件時の返り値)
    (停止条件でない時の処理)
```

例えば、ユークリッドの互助法を実装するときなどに再帰関数が使える<br>
$$
GCD(A, B) = GCD(B, A \% B) \, \text(B ≠ 0の時)
$$
$$
GCD(A, 1) = A \, \text(B = 0の時)
$$
ここで再帰関数を使うのは数式的にこうした定義になっているためである

In [14]:
def gcd(A, B):
    A, B = max(A, B), min(A, B) #AがBより大きいことを示す
    if B == 0:
        return A
    return gcd(B, A % B)
print(gcd(2,3))
print(gcd(2,4))
print(gcd(5,25))

1
2
5


他にも漸化式を実装するときなどに使える<br>
ここではフィボナッチ数列を実装してみよう<br>
$$
a_{n+2} = a_{n+1} + a_{n} \, \text(nが2より大きい時)
$$
$$
a_{n} = 1 \, \text(n = 0もしくはn = 1の時)
$$

In [18]:
def fibonacci(n):
    if n == 0 or n == 1:
        return 1
    return fibonacci(n-1) + fibonacci(n-2)
print(fibonacci(2))
print(fibonacci(10))

2
89


他にも再帰関数を使うケースはxのn乗の計算を以下のように考えて計算を削減する方法などがあります<br>
$$
x^n = (x^2)^{\frac{n}{2}} \, \text(nが偶数の時)
$$
$$
x^n = x(x^2)^{\frac{n-1}{2}} \, \text(nが奇数の時)
$$

ラムダ関数<br>
ラムダ関数は
```python
lambda (関数の引数): (関数の出力)
```
という形で書く関数である。<br>
lambdaを受けた変数は関数として振る舞う

In [7]:
pow_2 = lambda x: x**2
print(type(pow_2))
print(pow_2(10))
relu = lambda x: x if x > 0 else 0 #このようにif文を混ぜることもできる
print(relu(10))
print(relu(-10))
circle = lambda x,y: x**2 + y**2 #複数変数の場合はこのように定義する
print(circle(1,1))

<class 'function'>
100
10
0
2


ライブラリ<br>
ここいらでPythonという言語の中で最も優れている機能であるライブラリについて解説する。ライブラリはimport (ライブラリ名)でライブラリ<br>
を読み込み, (ライブラリ名).(ライブラリ内部で定義されている名前)でライブラリ内部の関数や変数を呼び出す。<br>

In [14]:
import math #mathライブラリは数学で使う定数などがまとめられたもの, 
print(math.pi) #mathライブラリの関数や定数を呼び出す時はmathを前につける
print(math.log(1))
from math import exp #from (ライブラリ) import (関数や変数)で読み込んだものはmathを前につけなくても動く
print(exp(1))
import math as m #ライブラリ名が長いときはこのように短縮した名前などをasで指定する
print(m.pow(2,3)) #pow関数はpow(x, y)の時xのy乗を計算する

3.141592653589793
0.0
2.718281828459045
8.0


Pythonに最初から入っているライブラリのことを標準ライブラリと呼ぶ。<br>
実は、それだけでもかなり多い[Python標準ライブラリ](https://docs.python.org/ja/3/library/index.html)<br>
このため、実際にプログラミングを行うときはChatGPTやGoogle検索なんかに<br>
```ChatGPT
Pythonで自動でメールを送るプログラムを作るコードを書いて
```
などと質問し、そこで出てきたライブラリを使うのがゼロから作るよりも早かったりする。<br>
もちろん、機械学習なんかもscikit-learnやPytorch, Tensorflowなどの外部ライブラリがあるので、[ゼロから](https://www.amazon.co.jp/ゼロから作るDeep-Learning-―Pythonで学ぶディープラーニングの理論と実装-斎藤-康毅/dp/4873117585/ref=asc_df_4873117585/?tag=jpgo-22&linkCode=df0&hvadid=295686767484&hvpos=&hvnetw=g&hvrand=16892698219903455766&hvpone=&hvptwo=&hvqmt=&hvdev=c&hvdvcmdl=&hvlocint=&hvlocphy=1009279&hvtargid=pla-526224399321&psc=1&th=1&psc=1)作る必要はない(しかしもちろんそうしたライブラリの中身を理解することは大切)<br>
外部ライブラリを利用したい場合は, コマンドラインで<br>
```bash
pip install (library name)
```
と書けば良い。pipは[PiPY(Python Package Index)](https://pypi.org)リポジトリにアクセスし、ライブラリを環境にインストールするコマンドである。<br>


しかし、PythonをいじっているとPiPYには登録されていないけどGithubにあるコードを使いたくなる時がある。<br>
今回はそうした時に使えるようにPythonのライブラリを簡単に学んでおこう

まずはターミナルを開き, cdコマンドなどでこのipynbファイルがあるフォルダに移動してこよう(詳しいことは調べるか発表者に聞こう)
```zsh
mkdir mylib
touch mylib/printhello.py
```
と入力しよう。(上のコマンドはMac, Linuxのみで動作します。WindowsのPowershellだとそれぞれmkdir, New-Itemあたりのコマンドが対応します。)<br>
制作したprinthello.pyには
```python
PI = 3.14159265358979
def printhello():
    print("Hello!")
```
とだけ書き込んでおこう。<br>
これを行った後次のコマンドを実行しよう

In [16]:
import mylib.printhello as printhello #ディレクトリの中にあるpyファイルは(ディレクトリ名).(ファイル名)のように指定する
printhello.printhello()
print(printhello.PI)

Hello!
3.14159265358979


In [17]:
from mylib.printhello import printhello, PI #これはmylib/printhello.py内部にあるprinthello関数とPIのみを取り出してくることを意味する
printhello()
print(PI)

Hello!
3.14159265358979


このように、作業しているディレクトリにあるpyファイルやディレクトリはライブラリとして読み込むことができる<br>
gitなどにあるコードをライブラリとして使用する場合は作業場所にcloneして使えば良い<br>
<br>
では, pipでインストールされたライブラリは作業しているディレクトリにないのに何故読み込めるのだろうか<br>
実はライブラリを読み込む際はPythonはPATHと呼ばれるものの中にあるライブラリを参照している。<br>
下のコードを実行してどこに実行している環境のPythonのPATHが通っているのかを確認しよう。

In [ ]:
import sys
print(sys.path) #個人情報なので実行したらgithubに上げる前に消すこと

#### クラス
一番重要な場所です。「クラスってなんですか?」と問われると, 説明が本一冊ぐらいになるらしいのですが(オブジェクト指向というプログラミング上での思想がある)<br>
とりあえずここでは便利な関数や数値を一まとめにしたものとして考えましょう。

クラスの定義は以下のテンプレートで行います。<br>
```python
class (クラス名):
    def __init__(self, (他の初期化引数)):
        ...
    
    def (クラスが持っていて欲しい関数):
        ...
    
```
というふうに実装します。<br>
<br>
このクラスというものはPythonに限らずJavaScriptやC++などのオブジェクト指向言語ならば必ず登場します。<br>
また, オブジェクト指向言語で作られたライブラリやフレームワークを使うときはこのクラスを理解していないがためにうまく扱えないことがほとんどです。<br>
これを機に覚えてしまいましょう

クラスはまず初期化の際に__init__関数が実行されます。__init__関数の引数には必ずselfを含んでください。<br>
selfは自分自身を意味するものであり、クラス内部の関数やメンバ変数をクラス内部で扱うときに使います(詳しくは後述)<br>

まずは組んでみましょう。class Personはname, ageという二つのプロバティを持ち, greetingsという関数(クラス内で定義される関数はメソッドと呼ぶ)を持ちます。<br>

In [22]:
class Person:
    def __init__(self, name: str, age: int)->None:
        self.name = name
        self.age = age
    def greetings(self)->None:
        print(f"My name is {self.name}. I am {self.age}") #

上のコードの解説をします。class Personが最初に実行される時, __init__関数がまず最初に実行されます。<br>
この時, 上のコードではself, name, ageが引数に入っていますが, selfは自動で入るためnameとageをクラスの最初の実行の際には引数に入れましょう。<br>
__init__やgreetingsのクラス内部の関数の最初の引数は必ずselfにしておきましょう。(お作法的な意味&事故を防ぐために)<br>
<br>
まずinitが実行されるとクラス自身を表すselfのself.nameとself.ageにクラスが最初に実行される時の引数のnameとageが代入されます。<br>
これらの引数にはclass Personの定義語にアクセスが可能です(下のコード参照)<br>
<br>
また, classで定義された__init__以外の関数には関数名のあとに()をつけることで呼び出しができます。<br>
クラス内部で__init__で定義された変数を使いたい場合には, 上のコードのように関数の引数にselfを加えて<br>
(例えばgreetings(self)のように), self.nameやself.ageのようにして呼び出して使います。<br>

In [23]:
person = Person("Tanaka", 19)
print(type(person))
print(person.name) #initで代入されたnameはperson.nameに表示される
print(person.age)
person.greetings() 

<class '__main__.Person'>
Tanaka
19
My name is Tanaka. I am 19


初めて学ぶ人にはクラスを学ぶ意味は薄く思えてしまうかもしれません(関数でよくないか?と思うかもしれません)<br>
しかし, Pythonで実際にみられるコードはクラスを使ったものがほとんどであり, <br>
Pythonを理解して他人のコードを見るためにクラスの理解は必要不可欠です。今はわからなくてもしっかり学んでおきましょう。

Tensorflow, Pytorchの例<br>
<img src="https://img.logmi.jp/article_images/AWd79enqJtPvJQbhyZ9bTF.png" width=75% height=75%>

Djangoの例<br>
<img src="https://i.stack.imgur.com/hO7Gq.png" width=75% height=75%>

クラスについてもう少し進んだ理解をしましょうPythonで実装されているもののほとんどは実はクラスです。<br>

In [24]:
print(type(list()))
print(type(dict()))
print(type(range(0)))
print(type(1))
print(type(1.0))
print(type(1+1j))

<class 'list'>
<class 'dict'>
<class 'range'>
<class 'int'>
<class 'float'>
<class 'complex'>


今まで紹介してきたリスト, 辞書, rangeなども, そしてint, float, complexなども<br>
全てPythonではクラスとして実装されています。<br>


以下のコードは先の辞書の解説に使ったコードです。このコードにおいて、<br>
keys(), values(), items()は辞書クラスで実装されているメンバ関数(メソッド)として解釈することができます。<br>

In [25]:
dictionary = {"Bob":16, "Mary":18, "Ken": 17}
print(dictionary["Bob"]) #dict[key]でそのkeyのvalueを確認できる
print(dictionary.keys()) #.keysメソッドでkeyを確認できる
print(dictionary.values()) #.valuesメソッドでvalueを確認できる
print(dictionary.items()) #.itemsで(key, value)のtupleで辞書の中身を確認できる
#上で述べたkeys, values, itemsは後に解説するfor文で中身を取り出すときに使われる

16
dict_keys(['Bob', 'Mary', 'Ken'])
dict_values([16, 18, 17])
dict_items([('Bob', 16), ('Mary', 18), ('Ken', 17)])


##### 特殊メソッド
クラスのより深い内容に踏み込みます。<br>
いくつかの特殊メソッド(メソッドとはクラス内部で定義される関数のことです)を紹介します。<br>
ステップバイステップで数学におけるベクトルを定義しましょう。
$$
v \in \mathbb{R^n} \, \, v = [v_{1}, v_{2}, ..., v_{n}]
$$

#### エラー文

#### 有名ライブラリ

#### Python環境構築
[【Python】M1/M2 MacでPython環境の構築（pyenvとvenvのインストールと設定、そして使い方）](https://namileriblog.com/python/python_pyenv_venv/)